src/foam/nanos/auth/UserPropertyAvailabilityService.js

```diff
@@ -23,30 +23,36 @@ foam.CLASS({
   ],
 
   javaImports: [
-    'foam.dao.ArraySink',
     'foam.dao.DAO',
-    'foam.mlang.MLang',
-    'foam.mlang.MLang.*',
-    'foam.nanos.auth.User'
+    'foam.nanos.auth.User',
+    'foam.nanos.theme.Theme',
+    'foam.nanos.theme.Themes',
+    'static foam.mlang.MLang.*'
   ],
 
   methods: [
     {
       name: 'checkAvailability',
       javaCode: `
-        DAO userDAO = (DAO) getX().get("localUserDAO");
-
-        if ( ! targetProperty.equals("userName") && ! targetProperty.equals("email") ) {
+        if ( ! targetProperty.equals("userName") &&
+             ! targetProperty.equals("email")
+        ) {
           throw new AuthorizationException();
         }
-
-        ArraySink select = (ArraySink) userDAO.inX(x).where(MLang.EQ(targetProperty.equals("userName") ? User.USER_NAME : User.EMAIL, value)).select(new ArraySink());
-        
-        if ( select.getArray().size() != 0 ) {
+        Theme theme = (Theme) ((Themes) x.get("themes")).findTheme(x);
+        if ( targetProperty.equals("email") && theme.getAllowDuplicateEmails() ) {
+          return true;
+        }
+        DAO userDAO = ((DAO) getX().get("localUserDAO")).inX(x);
+        User user = (User) userDAO
+          .find(AND(
+            EQ(User.getOwnClassInfo().getAxiomByName(targetProperty), value),
+            EQ(User.SPID, theme.getSpid())));
+        if ( user != null ) {
```

Comment:
next 4 lines could just be: return user == null;

Reply:
Right! thank you.